staker/block_validator.go

```diff
@@ -476,62 +476,40 @@ func (v *BlockValidator) validate(ctx context.Context, validationStatus *validat
 		"starting validation for block", "blockNr", entry.BlockNumber,
 		"blockAge", common.PrettyAge(time.Unix(int64(entry.BlockHeader.Time), 0)),
 		"blockDate", time.Unix(int64(entry.BlockHeader.Time), 0))
+	input, err := entry.ToInput()
+	if err != nil {
+		v.possiblyFatal(err)
+		return
+	}
+	gsExpected, err := entry.expectedEnd()
+	if err != nil {
+		v.possiblyFatal(err)
+		return
+	}
 	for _, moduleRoot := range validationStatus.ModuleRoots {
-
-		type replay = func(context.Context, *validator.ValidationInput, common.Hash) (validator.GoGlobalState, error)
-
-		execValidation := func(replay replay, validationType string) error {
-			input, err := entry.ToInput()
-			if err != nil {
-				return err
+		before := time.Now()
+		for _, spawner := range v.validationSpawners {
+			gsEnd, err := spawner.Execute(ctx, input, moduleRoot)
+			if err == nil && gsEnd != gsExpected {
+				err = fmt.Errorf("unexpected GlobalState. Got: %v", gsEnd)
```

Comment:
Should we log what global state we expected here?

Reply:
added